Pipeline Editor: Allow for configuration of shared memory size

ptitzler opened the pull request:
~Requires https://github.com/elyra-ai/elyra/pull/2957~ RESOLVED

Extend the pipeline editor to allow for setting a custom shared memory size in generic and custom components. A pipeline default can be configured in the pipeline properties. Support is limited to Kubeflow Pipelines and Apache Airflow. 

![image](https://user-images.githubusercontent.com/13068832/195720297-c73d1229-1f6a-4a32-9d7a-d15325cd0ee4.png)

To define a custom size the user needs to specify a value other than zero. Zero implies that the Kubernetes default is used, which is 64Mi.

![image](https://user-images.githubusercontent.com/13068832/194901254-0cfd4ac7-4927-442a-8025-8b01b78a8b0b.png)



When a custom shared memory size is configured, Elyra adds a volume and volume mount to the pod, as shown here for Kubeflow Pipelines:
```
    volumeMounts:
     - {mountPath: /dev/shm, name: shm}
    volumes:
    - emptyDir: {medium: Memory, sizeLimit: 2G}
      name: shm
```
Closes #2838

TODO:

- [x] pipeline editor release and merge

<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our [contributor guidelines](https://github.com/elyra-ai/community/blob/main/contributing.md)
    1.1 Please follow the [7 rules for a great commit message](https://github.com/elyra-ai/community/blob/main/contributing.md#creating-a-pull-request)
  2. If the PR is unfinished, leave it as 'DRAFT', add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...' or use the 'status:Work in Progress' label.
  3. Be sure to keep the PR description updated to reflect all changes.
  4. Please write your PR title to summarize what this PR proposes.
  5. If this PR involves UI changes, please provide a screen capture (before/after) for a faster review.
  6. Remember to add 'Fixes #ISSUE_NUMBER' (or any [valid keyword](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)) to the end of your message body to get the issue properly closed when the PR is merged.
  7. Set the proper milestone based on the target release for the issue.
  8. Consider whether any documentation need to be added or updated based on your changes.
-->

### What changes were proposed in this pull request?
- See above
- Add new property description to pipeline documentation
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue.
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor code that leads to class changes, showing the updated class hierarchy will help reviewers.
  2. If there is design documentation, please add the link.
-->

### How was this pull request tested?

- Added validation tests (for valid and invalid inputs)
- Updated handlers test

pipeline defaults:

- [x] no custom size specified
- [x] invalid custom size specified
- [x] valid size specified

generic nodes (KFP and Airflow):

- [x] no custom size specified and no pipeline default defined
- [x] invalid custom size specified
- [x] valid size specified

custom nodes (KFP and Airflow):

- [x] no custom size specified and no pipeline default defined
- [x] invalid custom size specified
- [x] valid size specified

<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly
including negative and positive cases if possible.

If it was tested in a way different from regular unit tests, please clarify how you tested step by step,
ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.

If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

 
Developer's Certificate of Origin 1.1

       By making a contribution to this project, I certify that:

       (a) The contribution was created in whole or in part by me and I
           have the right to submit it under the Apache License 2.0; or

       (b) The contribution is based upon previous work that, to the best
           of my knowledge, is covered under an appropriate open source
           license and I have the right under that license to submit that
           work with modifications, whether created in whole or in part
           by me, under the same open source license (unless I am
           permitted to submit under a different license), as indicated
           in the file; or

       (c) The contribution was provided directly to me by some other
           person who certified (a), (b) or (c) and I have not modified
           it.

       (d) I understand and agree that this project and the contribution
           are public and that a record of the contribution (including all
           personal information I submit with it, including my sign-off) is
           maintained indefinitely and may be redistributed consistent with
           this project or the open source license(s) involved.

marthacryan commented:
Added a couple updates - now showing the default value as a placeholder, updated the UI for object fields to look a little clearer. 
![image](https://user-images.githubusercontent.com/6673460/192901924-5a7f5ca7-67ec-456a-be82-4372af524e47.png)
![image](https://user-images.githubusercontent.com/6673460/192901938-8eeecfd4-02e1-4009-885d-a3937a975eea.png)

ptitzler commented:
> **Note**: now requires a pipeline-editor release. Added a couple updates - now showing the default value as a placeholder, updated the UI for object fields to look a little clearer.

Visually this is looking great! Thank you! I did notice two things:
 - It's possible to enter `e` or `E` as value, which is stored as -66. Validation catches it though but it's odd that it's possible that this letter can be entered but no others. My best guess is that this has to do with the [e-notation](https://en.wikipedia.org/wiki/Scientific_notation#E_notation).
   
   ![image](https://user-images.githubusercontent.com/13068832/192946479-128a4088-fec6-4c07-b475-a1aa2d3124ad.png)

 - The pipeline default value isn't rendered in the node properties view. Based on your description I assumed it would be rendered as the placeholder (size) and pre-selection (units). In the image below `0`/`Mi` is displayed instead of `13`/`Gi` Can you please clarify?
![image](https://user-images.githubusercontent.com/13068832/192947615-f5e79372-966b-4e6e-a152-da8b172080e1.png)

ptitzler commented:
In today's community meeting it was brought up that giving the user the option to choose from multiple units might be complicating things. While I agree that it might not be necessary to support Gi/Mi and G/M by choosing a basis that is applied to all memory units in Elyra, there would benefits to giving the user the option to choose a unit (e.g GB/MB ignoring for now what basis is used) _if the units are supposed to be internally interpreted using power of two_ instead of power of ten. 

It's not clear to me which basis is used today for the generic components RAM resources setting. The implementation for the KFP and Airflow processor appear to be in disagreement, with the KFP processor applying the G (power of ten-based) suffix, whereas the Airflow processor applying none (which may or may not be a bug). Anybody have insights whether G or Gi is the "correct" one?

ptitzler commented:
Deferring to next release to avoid delaying 3.12 further.

ptitzler commented:
Follow-up required:  When I merged the `main` branch into this PR branch I overrode the pipeline editor dependency because I believed the necessary changes were already included  already included the the most recent pipeline editor release. I'm not sure if this assumption is true. @marthacryan is https://github.com/elyra-ai/pipeline-editor/pull/202 specific to this PR?

ptitzler commented:
> Follow-up required: When I merged the `main` branch into this PR branch I overrode the pipeline editor dependency because I believed the necessary changes were already included already included the the most recent pipeline editor release. I'm not sure if this assumption is true. @marthacryan is [elyra-ai/pipeline-editor#202](https://github.com/elyra-ai/pipeline-editor/pull/202) specific to this PR?

Discussed this during today's dev meeting. A link to the PR's sandbox link needs to be re-added

akchinSTC commented on elyra/pipeline/airflow/processor_airflow.py:
```diff
@@ -723,6 +750,7 @@ def supported_properties(self) -> Set[str]:
             pipeline_constants.KUBERNETES_POD_ANNOTATIONS,
             pipeline_constants.KUBERNETES_POD_LABELS,
             pipeline_constants.KUBERNETES_TOLERATIONS,
+            pipeline_constants.KUBERNETES_SHARED_MEM_SIZE,
         ]
```
```suggestion
            pipeline_constants.KUBERNETES_SHARED_MEM_SIZE,
        }
```
Should be curly braces for sets

akchinSTC commented on elyra/pipeline/airflow/processor_airflow.py:
```diff
@@ -597,6 +598,12 @@ def render_volumes(self, elyra_parameters: Dict[str, ElyraProperty]) -> str:
                 Volume(name="{v.pvc_name}", configs={{"persistentVolumeClaim": {{"claimName": "{v.pvc_name}"}}}}),"""
+        # set custom shared memory size
+        shm = elyra_parameters.get(pipeline_constants.KUBERNETES_SHARED_MEM_SIZE)
+        if shm is not None and shm.size:
+            config = f"""configs={{"emptyDir": {{"medium": "Memory", "sizeLimit": "{shm.size}{shm.units}"}}}}"""
```
`shm.size` and `shm.units` doesn't look like they are defined in the `ElyraProperty` ABC.

kevin-bates replied:
These come from the object-valued property subclass: `CustomSharedMemorySize`.

akchinSTC commented on elyra/pipeline/component_parameter.py:
```diff
@@ -298,6 +299,100 @@ def add_to_execution_object(self, runtime_processor: RuntimePipelineProcessor, e
+
+    def get_value_for_dict_entry(self) -> str:
+        """Returns the value to be used when constructing a dict from a list of classes."""
+        return self.to_dict()
```
```suggestion
        return self.to_dict()
```
Returns a dict but typehints to a str, should this be Any if the return value can be both str and dict?

kevin-bates replied:
This method is on `ElyraPropertyListItem` and its subclasses, yet `CustomSharedMemorySize` is a subclass of `ElyraProperty` (and sibling to `ElyraPropertyListItem`).  What is going to call `get_value_for_dict_entry()` on `CustomSharedMemorySize` instances?

Same goes for `to_dict()` - this is only defined for classes of `ElyraPropertyListItem`.

Is this class deriving from the correct superclass or can these two methods simply go away?

Should it be a "list item" simply because it consists of two attributes (one of which is currently hidden)?

kiersten-stokes replied:
Good catch! After checking this out, I added a commit that addresses two things:

- parent `get_value_for_dict_entry` method needs to return `Union[str, Dict[str Any]]`
- the `CustomSharedMemorySize` class itself doesn't need either `get_value_for_dict_entry` or `get_key_for_dict_entry`, so these methods are removed

kiersten-stokes replied:
@kevin-bates ha! we must have hit respond at the same time. You're completely right though, these methods can go away!

akchinSTC commented on elyra/pipeline/kfp/processor_kfp.py:
```diff
@@ -833,6 +850,7 @@ def supported_properties(self) -> Set[str]:
             pipeline_constants.KUBERNETES_POD_LABELS,
             pipeline_constants.KUBERNETES_TOLERATIONS,
             pipeline_constants.DISABLE_NODE_CACHING,
+            pipeline_constants.KUBERNETES_SHARED_MEM_SIZE,
         ]
```
curly braces for sets

kevin-bates commented on elyra/pipeline/component_parameter.py:
```diff
@@ -298,6 +299,100 @@ def add_to_execution_object(self, runtime_processor: RuntimePipelineProcessor, e
+            hidden=False,
+            required=False,
+        ),
+        PropertyAttribute(
+            attribute_id="units",
+            display_name="Units",
+            input_type="string",
+            hidden=True,
+            required=False,
+        ),
```
Does this (and the corresponding validation logic) exist solely in case we decide to expose units someday?  Is there a precedent for this kind of "crystal ball" thinking? :smile:

If so, it seems like the `display_name` value of the _size_ (e.g., `display_name="Memory Size (GB)",`) should also "go through the motions" of conveying the (possibly) variable unit types as well.

ptitzler replied:
> exist solely in case we decide to expose units someday

Yep!

> Is there a precedent for this kind of "crystal ball" thinking

I'm not sure there is, but the main benefit is that this serves as a blueprint for the "hidden" property, which was introduced in support of this PR.

> If so, it seems like the display_name value of the size (e.g., display_name="Memory Size (GB)",) should also "go through the motions" of conveying the (possibly) variable unit types as well.

Since there are code changes required to make this property visible (here and in the tests, which for now treat other 'valid unit specifications'  as invalid [1]), I figured the same code change would update the display name by removing the units that are now selected by default and the documentation.

kevin-bates commented on elyra/pipeline/component_parameter.py:
```diff
@@ -378,7 +457,10 @@ def get_schema(cls) -> Dict[str, Any]:
         return schema
 
     def get_value_for_dict_entry(self) -> str:
```
We should also update the return value types of each override to that of the superclass: `Union[str, Dict[str, Any]]:`

kevin-bates reviewed (approved):
LGTM
